### DISCUSSION TITLE
post/3/

### DISCUSSION THREAD

**Discussion by giscus** [HoshinoRei/blog#2] [General]
# post/3/

在这篇文章里，小玲将教你如何将一个装着干净的 Debian 系统的设备变成一个能让局域网内设备上网的软路由。小玲使用一台 Nanopi R4S 作为演示用的设备，刷入的系统为 Armbian 21.08.1，Debian 版本为 11。你也可以用 X86 的设备，设置过程基本上没有区别。

https://blog.bling.moe/post/3/

**Comment by netcan** [comment]
请教一个问题，我的nftable默认将input流量drop掉，只允许lan口流量通过，这样会导致pppoe拨号拿不到公网ipv6地址，你知道原因么？是不是公网ipv6协商得开放某些特殊端口？
```
table inet filter
delete table inet filter
table inet filter {
        chain input {
                type filter hook input priority 0; policy drop;
                ct state established,related accept;
                iifname { enp1s0, lo } accept; // <- 加上ppp0可以拿到公网ipv6地址
                tcp dport 22 ct state new counter;
        }
        chain forward {
                type filter hook forward priority 0;
        }
        chain output {
                type filter hook output priority 0;
        }
}
```

**Reply by netcan** [reply]
看了下openwrt的配置，应该是要配dhcp-v6规则https://openwrt.org/docs/guide-user/network/ipv6/configuration#ppp-based_protocols_and_option_ipv6

**Comment by beyes** [comment]
“重启之后 R4S 的 LAN 口的 IP 地址已经变成了 192.168.3.0/24 这个网段了，不能再通过将 R4S 的 LAN 口连接原来的路由器的 LAN 口来获取原来的路由器的 IP 地址了。所以咱们需要使用一条网线连接着 R4S 的 LAN 口和电脑，一条网线连接的原本路由器的 LAN 口和 R4S 的 WAN 口。 ” 想问一下Lan 口接路由器开Ap 模式，Wlan 口接光猫不行吗？电脑没有网口？

**Reply by HoshinoRei** [reply]
行。

**Reply by beyes** [reply]
我明白你那样操作是让电脑仍然能连外网

**Comment by beyes** [comment]
请教一下
最后一步添加miniupnpd 功能 执行完 sudo docker-compose -f ~/docker-compose.yml up -d 提示：

services.wide-dhcpv6-client  Additional property miniupnpd is not allowed

这是什么原因呀，其他步骤都走通了？

**Reply by beyes** [reply]
我重启 docker 之后可以了

**Comment by beyes** [comment]
不清楚你那边是否有问题，我按照你的教程走完发现路由功能没问题，但是我在服务器上开下载发现本地的 /etc/resolv.conf 配置会被 dnsmasq 重写，重写的 dns 是无效的？ 网上查了一下是 dnsmasq 的bug。

**Reply by beyes** [reply]
Manual forwarding

First you must set localhost addresses as the only nameservers in /etc/resolv.conf:

/etc/resolv.conf

nameserver ::1
nameserver 127.0.0.1
options trust-ad

Make sure to protect /etc/resolv.conf from modification as described in Domain name resolution#Overwriting of /etc/resolv.conf.

The upstream DNS server addresses must then be specified in dnsmasq's configuration file as server=server_address. Also add no-resolv so dnsmasq does not needlessly read /etc/resolv.conf which only contains the localhost addresses of itself.

/etc/dnsmasq.conf

[...]
no-resolv

# Google's nameservers, for example
server=8.8.8.8
server=8.8.4.4

Now DNS queries will be resolved with dnsmasq, only checking external servers if it cannot answer the query from its cache.

**Reply by HoshinoRei** [reply]
没看懂你想问什么。下次问问题最好把问题描述清楚，不要让回答的人猜你想问什么。

1. 你的软路由用的什么设备？X86 还是 ARM？
2. 你的软路由装的系统是什么系统？是否纯净？
3. 你的软路由是不是就是你说的“服务器”？如果是，除了小玲的文章写的软件以外，你还装了什么？做了什么小玲的文章里没有提到的事？
4. 还是说你在软路由下面接了一台服务器，通过软路由的 DHCP 获取 IP。

**Reply by beyes** [reply]
我可能没描述清楚，直接贴了网上查到的文档。不好意思。我软路由用的 debian，x86 的 cpu，纯净系统，我的软路由就是我的服务器，我在 debian 上跑你文章中的路由服务，我在 debian 上还装了qbittorrent-nox 跑下载，以及 samba 服务在局域网中使用。我按照你文章的指导配置好了我的debian 服务器的路由功能，接上ap，连接ap上网用其他设备的 qbittorrent 下载，下载功能一切正常。debian 机器本身的 qbittorrent-nox 不能下载了，我发现/etc/resolv.conf会在重启后被重写，也不确定是哪个服务重写的，后面看文档，就把 /etc/resolv.conf 设置为不可更改，文件 dnsmasq.conf 加入no-resolv 后问题解决。

**Reply by HoshinoRei** [reply]
好吧，既然解决了就好。小玲也要告诉你小玲自己没有 `/etc/resolv.conf` 被重写的问题。

**Reply by beyes** [reply]
查找了好久，终于找到是哪个脚本在重写 /etc/resolv.conf，对于 Debian 11，我想其他人可能也会遇到类似的问题，我在这里写一下：dhclient-script 会重写 /etc/resolv.conf，我使用的方法是创建一个 hook 去防止 /etc/resolv.conf 被更新，具体来说编辑 vim /etc/dhcp/dhclient-enter-hooks.d/nodnsupdate 的内容为
 #!/bin/sh
make_resolv_conf(){
	:
}
并给它加上执行权限 chmod +x /etc/dhcp/dhclient-enter-hooks.d/nodnsupdate 这样 dhclient-script 中的 make_resolv_conf() 函数将被我们自己定义的覆盖。

另外，我分析你的没有出现这个情况是因为一开始你把 NetworkManager 服务删除了，按照这个思路 sudo vim /etc/NetworkManager/NetworkManager.conf 在[main] 中添加 dns=none 应该也可以。

**Comment by ByteJumper** [comment]
遇到个奇怪的问题。有空了帮忙指导下。
x86-64小主机通过网络netinst.iso新安装的amd-64的debian11.6系统做路由器。
配置了网卡、ipv4转发、dnsmasq、nftables和pppoeconf。
没安装 wide-dhcpv6-client 和 miniupnpd 。

第1种网络情况
光猫 → 拨号路由 → 小主机 → 电脑
小主机如果通过DHCP连接拨号路由器，小主机和通过小主机联网的电脑都正常，小主机和电脑ping ip或域名都可以通。小主机wget 也可以下载文件。电脑也能打开网页。

但是。如果小主机拨号做路由的话，出现奇怪的故障。
第2种网络情况
光猫 → 小主机 → 电脑
小主机负责拨号连外网和局域网路由。
小主机 拨号能成功，也能ping通域名和ip，也能wget下载文件。apt安装下载软件也正常。
但是，通过小主机连接的电脑就不行。
电脑能ssh连接小主机，也能ping通ip和域名，能登上qq，能打开部分网页，部分网页打不开（如淘宝、京东）。
能打开的网页是域名直接解析到服务器主机ip的。
不能打开的网页大都是使用了云加速的网站。
nslookup 和 traceroute 也没发现明显异常。
查看dnsmasq的log也没发现异常

题外话，原来我安装的debian11.5系统一样的配置，并且做了备份。不论小主机dhcp还是ppoe拨号都能正常联网，不存在上述情况。
但是由于换了硬盘，重新装了系统。新安装的系统是debian11.6，按照备份的配置，就出现上述奇怪的情况。

实在不知道什么原因，如果方便的话，请帮忙指导下排查思路，或者可能的问题在哪。
先谢谢了！

**Reply by HoshinoRei** [reply]
看一下你的 DNS 是否有问题，用运营商提供的 DNS 就好。

**Reply by ByteJumper** [reply]
不是DNS的问题。最后nftables配置文件加了转发
table ip mangle {
        chain FORWARD {
                type filter hook forward priority mangle; policy accept;
                oifname ppp0 tcp flags & (syn|rst) == syn tcp option maxseg size 1400-65495 counter tcp option maxseg size set rt mtu
        }
}
才解决上述问题。

**Comment by lazywen** [comment]
楼主请教下客户端成功获取到ipv6地址（240e:xxxx）后这个地址与外界不通，只有ppp0上的ipv6地址与外界是通的，是不是ipv6没做nat的关系，检查过防火墙了没有任何禁止的规则

**Reply by HoshinoRei** [reply]
请问"与外界不通"是指从内部访问外部还是外部访问内部？

**Reply by lazywen** [reply]
除了ppp0的ipv6正常以外，从内网ping6其它设备获取的ipv6地址不通，当然从外网ping6也不通，虽然看起来是公网ipv6地址，访问ipv6-test网站测试也显示没有ipv6地址，就没搞明白为什么楼主没添加ipv6 nat规则也能使用，因为ipv4是做了转换的，但是试了添加了ipv6 nat规则也还是不行，还没找到是哪的原因

**Reply by lazywen** [reply]
我照搬这个配置，问题解决了不劳烦博主了
https://github.com/jinmei/wide-dhcpv6/blob/master/dhcp6c.conf.sample

**Comment by tangzhi1996** [comment]
我只想说,我也遇上这样的需求了,入手了一个ARM(H68K),四网口的那种,我想拿他看电视,当路由,开AP,配置,配得我头痛

**Comment by felixxiao99** [comment]
请问如果要在借助armbian进行科学上网需要怎么操作，大部分使用软路由的目的都是这😬

**Comment by shilazi** [comment]
你的文章真的很有帮助，写得非常棒！！！

我遇到一个搞不定也不理解的问题，是 ipv6 地址到期后没有新的地址可用。

### 软硬件说明

* J4125 x86 四口迷你主机
* Debian 12 bookworm 纯净系统
* 奠信 300M 宽带

### 操作说明

总体按照博文的步骤操作，结合自身实际情况做配置修改。

### 不同之处

* 操作系统用的 Debian 12

```
root@j4125:~# cat /etc/debian_version
12.2

root@j4125:~# uname -a
Linux j4125 6.1.0-13-amd64 #1 ebian 6.1.55-1 (2023-09-29) x86_64 GNU/Linux
```

* 修改拨号 WAN 接口名称

```
auto wan0
iface wan0 inet ppp
pre-up ip link set enp1s0 up
provider wan0
```

* 使用桥接接口 LAN 侧接口

```
auto lan0
iface lan0 inet static
  address 192.168.233.1/24
  bridge_ports enp2s0 enp3s0 enp4s0
  bridge_maxwait 0
  bridge_fd 0
```

* 自定义 dhcp6c.conf 配置（因为奠信给的 PD 是 56）

```
interface wan0 {
    send ia-pd 0;
};
id-assoc pd 0 {
    prefix-interface lan0 {
        sla-id 233;
        sla-len 8;
        ifid 1;
    };
};
```

* ipv6 相关 sysctl 配置引用的 [X86 软路由配置 IPv6 踩坑小记](https://www.v2ex.com/t/724060)

```
net.ipv6.conf.all.forwarding=2
net.ipv6.conf.default.forwarding=2

net.ipv6.conf.all.accept_ra=2
net.ipv6.conf.default.accept_ra=2

net.ipv6.conf.all.use_tempaddr=2
net.ipv6.conf.default.use_tempaddr=2
```

### 配置好之后的表现

* 拨号成功后，wan0 接口能正确获取到 ipv6 的 EUI64 和临时地址

```
1: wan0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1492 state UNKNOWN group default qlen 3
    link/ppp
    inet6 240e:xxxx:88b2/64 scope global temporary dynamic
       valid_lft 258232sec preferred_lft 82753sec
    inet6 240e:xxxx:9746/64 scope global dynamic mngtmpaddr
       valid_lft 258232sec preferred_lft 171832sec
    inet6 fe80::9746 peer fe80::9f90/128 scope link
       valid_lft forever preferred_lft forever
```

* 启动 wide-dhcpv6-client 后，lan0 接口能正确获取到 ipv6 的 EUI64 和临时地址。这里就产生**第一个疑问**，你的博文里，lan 口只生成了 ::/64 的 ipv6 地址，并没生成 EUI64 和临时地址。**第二个疑问**是在 EUI64 地址的 valid_lft 和 preferred_lft 读秒结束时，没有新的 ipv6 地址生成（表现为没有续期（不知道表述对不对）或新增新地址）

```
2: lan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP group default qlen 1000
    inet 192.168.233.1/24 brd 192.168.233.255 scope global lan0
       valid_lft forever preferred_lft forever
    inet6 240e:xxxx:dff9/64 scope global temporary dynamic
       valid_lft 41380sec preferred_lft 41380sec
    inet6 240e:xxxx:6a81/64 scope global dynamic mngtmpaddr
       valid_lft 41380sec preferred_lft 41380sec
    inet6 240e:xxxx::1/64 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::6a81/64 scope link
       valid_lft forever preferred_lft forever
```

### 其他拓展

我使用 TP-link 的路由器插入 lan0 接口，并开启桥接模式。现在是一台笔记本接入 WIFI，使用 VMware 开了一台 Debian 12 虚拟机，网络使用的桥接，/etc/network/interface 未做任何修改，/etc/sysctl.conf 未做任何修改，然后获取到的 ipv6 的 EUI64 地址。面临跟 lan0 接口 ipv6 一样的情况，在 EUI64 地址的 valid_lft 和 preferred_lft 读秒结束时，inet6 地址直接消失，大概过了 3 到 5 分钟，又会重新添加 inet6 地址，我在想是不是因为 lease_time 不能设置太短。

```
2: ens33: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 state UP group default qlen 1000
    inet 192.168.233.114/24 brd 192.168.233.255 scope global dynamic ens33
       valid_lft 17740sec preferred_lft 17740sec
    inet6 240e:xxxx:f03e/64 scope global dynamic mngtmpaddr
       valid_lft 40672sec preferred_lft 40672sec
    inet6 fe80::f03e/64 scope link
       valid_lft forever preferred_lft forever
```

### 写在最后

只是把自己的配置贴出来，看看你能不能给些建议或者方向，这样我就能按症下药了，先行感谢。

**Reply by HoshinoRei** [reply]
小玲以前用 Debian 11 时，也不知怎么回事，LAN 口没有 EUI64 地址和临时地址。后来升级到了 Debian 12 后好像开始有了。具体有没有小玲已经不记得了，因为小玲现在采用 [NetworkManager](https://blog.bling.moe/post/19/) 了，`wide-dhcpv6-client` 也不用了。现在前缀的自动更新是没问题的。

```text
2: enp1s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq state UP group default qlen 1000
    link/ether ██:██:██:██:██:██ brd ff:ff:ff:ff:ff:ff
    inet 192.168.3.1/26 brd 192.168.3.63 scope global noprefixroute enp1s0
       valid_lft forever preferred_lft forever
    inet6 240█:████:████:████::1/64 scope global dynamic noprefixroute
       valid_lft 238306sec preferred_lft 151906sec
    inet6 fe80:::████::████::████::████/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
```

**Reply by shilazi** [reply]
感谢小玲的回复，至少有了可能的解决方向。

**Reply by shilazi** [reply]
使用 [radvd](https://tldp.org/HOWTO/Linux+IPv6-HOWTO/ch22s04.html) 之后续期正常，细看日志像是 [dnsmasq](https://wiki.archlinux.org/title/dnsmasq) 忽略了 `client`（限定 `win10`、`debin12`） 的 `RS` 请求。

### radvd

会收到来自 `windows`、`linux` 的 `RS` 请求。

```
[Dec 10 02:24:34] radvd (7): lan0 recvmsg len=16
[Dec 10 02:24:34] radvd (7): lan0 received a packet
[Dec 10 02:24:34] radvd (7): lan0 received RS from: fe80::xxxx:f03e
[Dec 10 02:24:34] radvd (7): ::/64 auto prefix detected on iface lan0
[Dec 10 02:24:34] radvd (7): auto-selected prefix 240e:xxxx::/64 on interface lan0
[Dec 10 02:24:34] radvd (7): sending RA to fe80::xxxx:f03e on lan0 (fe80::xxxx:6a81), 3 options
[Dec 10 02:24:34] radvd (7): lan0 processed an RS
```

### dnsmasq

**仅** 会收到来自 `windows` 的 `RS` 请求。

```
Dec 10 10:41:12 dnsmasq-dhcp[10044]: RTR-ADVERT(lan0) 240e:xxx::
Dec 10 10:41:30 dnsmasq-dhcp[10044]: RTR-SOLICIT(lan0) c6:xx:xx:0e
Dec 10 10:45:57 dnsmasq-dhcp[10044]: RTR-ADVERT(lan0) 240e:xxx::
```


上面提到的 **大概过了 3 到 5 分钟，又会重新添加 inet6 地址** 应该是 [dnsmasq](https://wiki.archlinux.org/title/dnsmasq) 的新一轮 `RA` 事件发布了。

推测原因是 [dnsmasq](https://wiki.archlinux.org/title/dnsmasq) 无法响应 `RS`，因此 `client` 需根据 `RA` 事件配置 `slaac ipv6`。

自问自答给出两个解决方案，以下两者各有利弊，供君参考。

1. 不使用 [dnsmasq](https://wiki.archlinux.org/title/dnsmasq) 的 `enable-ra` 配置项，`RA` 事件通过 [radvd](https://tldp.org/HOWTO/Linux+IPv6-HOWTO/ch22s04.html) 通告。
2. 缩减 [dnsmasq](https://wiki.archlinux.org/title/dnsmasq) `RA` 事件通告的间隔，即提高通告的频率，设置参数 `ra-param`。

**Comment by TurnOffNOD** [comment]
咱博客的文章，post/9 和 post/4 在哪里了？

**Reply by HoshinoRei** [reply]
回归虚无力。

**Reply by TurnOffNOD** [reply]
？？？啥意思？

**Reply by HoshinoRei** [reply]
就是没有了。

**Comment by Endt4sk** [comment]
我在debian上翻墙之后，别的机器可以正常访问google.com，就是访问不了github.com，但是在debian上访问github.com又是正常的，不知道是怎么回事
